pkg/endpoint/endpoint.go

```diff
@@ -1263,40 +1225,139 @@ func (e *Endpoint) GetDockerNetworkID() string {
 
 // GetState returns the endpoint's state
 // endpoint.Mutex may only be RLock()ed
+func (e *Endpoint) GetStateLocked() string {
+	return e.state
+}
+
+// GetState returns the endpoint's state
+// endpoint.Mutex may only be RLock()ed
 func (e *Endpoint) GetState() string {
 	e.Mutex.RLock()
 	defer e.Mutex.RUnlock()
-	return e.State
+	return e.GetStateLocked()
 }
 
-// SetState modifies the endpoint's state
-// endpoint.Mutex may not be held
-func (e *Endpoint) SetState(state string) {
-	e.Mutex.Lock()
-	e.State = state
-	e.Mutex.Unlock()
+// SetStateLocked modifies the endpoint's state
+// endpoint.Mutex must be held
+// Returns true only if endpoints state was changed as requested
+func (e *Endpoint) SetStateLocked(toState string) bool {
```

Comment:
Would be nice to have this method unit tested to ensure that all transitions are correct. Unit test would also act as a documentation on what transitions are allowed.

Reply:
Good point. I've added a full matrix as well as state transitions during a typical lifecycle of an endpoint.